### PR TITLE
Add the possibility to run Dynalab on GPU

### PR DESCRIPTION
**Pull request by gwenzek**
For now I've hardcoded to only run Flores tasks on GPU, but we could imagine exposing this as a flag to the user.

I've tested on a "Deep Learning AMI (Ubuntu 18.04) Version 44.0" AWS instance

### CODE REVIEW

File: dynalab_cli/test.py [hunk 1]
```diff
@@ -64,132 +83,155 @@ def run_command(self):
...
-            )
-            docker_path = os.path.join(tmp_dir, "Dockerfile")
-            # TODO: pull the files from dynalab repo once public
-            shutil.copyfile(os.path.join(lib_dir, "Dockerfile.dev"), docker_path)
-            shutil.copyfile(
-                os.path.join(lib_dir, "dev-docker-entrypoint.sh"),
-                os.path.join(tmp_dir, "dev-docker-entrypoint.sh"),
+            self.run_docker_test(config)
+
+    def run_docker_test(self, config):
```
Comment (gwenzek) [inline dynalab_cli/test.py:89]: this diff is pretty bad. I just split the `run_command` in two functions, and de-indented.

File: dynalab_cli/test.py [hunk 2]
```diff
@@ -64,132 +83,155 @@ def run_command(self):
...
+        lib_dir = os.path.join(
+            os.path.dirname(os.path.dirname(os.path.realpath(__file__))),
+            "dynalab",
+            "dockerfiles",
+        )
+
+        use_gpu = self.use_gpu(config)
+        docker_file = Path(lib_dir) / "Dockerfile.dev"
+        if use_gpu:
+            docker_file = Path(lib_dir) / "Dockerfile.cuda"
```
Comment (gwenzek) [inline dynalab_cli/test.py:152]: Here are some actual changes

File: dynalab_cli/test.py [hunk 3]
```diff
@@ -64,132 +83,155 @@ def run_command(self):
...
+            repository_name,
+            "-f",
+            docker_path,
+            ".",
+        ] + docker_build_args
+
+        subprocess.run(docker_build_command)
+        # NOTE: cpu and memory limit are specific to ml.m5.xlarge
+        # breakpoint()
+        docker_run = [
```
Comment (gwenzek) [inline dynalab_cli/test.py:188]: and here

File: dynalab_cli/utils.py [hunk 4]
```diff
@@ -98,6 +98,7 @@ def login():
             "A browser window will open prompting you to login, "
             + "after login please copy the API token and paste it here"
         )
+        print(f"Opening in browser: {DYNABENCH_WEB}/generate_api_token")
```
Comment (gwenzek) [inline dynalab_cli/utils.py:101]: this is useful if you're running dynalab on a server

File: dynalab_cli/utils.py [hunk 5]
```diff
@@ -146,11 +147,11 @@ def default_filename(key):
 
 
 def check_model_name(name):
-    pat = re.compile("^[a-zA-Z0-9-]+$")
+    pat = re.compile("^[a-z0-9-]+$")
```
Comment (gwenzek) [inline dynalab_cli/utils.py:150]: Sagemaker jobs uses the model name and need to be lowercased.

**Review by douwekiela — APPROVED** [review]
LGTM

File: dynalab_cli/test.py [hunk 6]
```diff
@@ -33,6 +36,22 @@ def __init__(self, args):
         self.args = args
         self.config_handler = SetupConfigHandler(args.name)
 
+    def use_gpu(self, config) -> bool:
+        task = config["task"]
+        # TODO: should this be stored in the DB ? And exposed in the /tasks API ?
```
Comment (mazhiyi) [inline dynalab_cli/test.py:41]: There's no db in dynalab. Can write a task config in the e.g. common.py file instead

File: dynalab_cli/utils.py (hunk 5 above)
Comment (mazhiyi) [reply dynalab_cli/utils.py:150]: I think the model name will be lower cased when creating a job.

Comment (mazhiyi) [reply dynalab_cli/utils.py:150]: I don't think sagemaker has that requirements. We've got a couple of job names with upper case

File: dynalab_cli/test.py [hunk 7]
```diff
@@ -64,132 +83,154 @@ def run_command(self):
...
-                docker_path,
-                ".",
-            ] + docker_build_args
-
-            subprocess.run(docker_build_command)
-            # NOTE: cpu and memory limit are specific to ml.m5.xlarge
-            process = subprocess.run(
-                f"docker run --network none --cpus 4 --memory 16G {repository_name}",
-                shell=True,
-                stderr=subprocess.PIPE,
```
Comment (mazhiyi) [inline dynalab_cli/test.py:165]: The reason we moved all torchserve logs into stderr, and print out stdout here is that, during integrated test, we want people to see the test data and test response on the console (i.e. the printouts in the docker entrypoint file), but redirect all the torchserve logs into the file for later inspection so it doesn't crowd the console. 

Can we resume to the previous design please?

**Pull request merged**